Update EIP-5496: Change status to Last Call (#5768)
ethereum/EIPs: EIPS/eip-5496.md

---
eip: 5496
title: Multi-privilege Management NFT Extension
description: Create shareable multi-privilege NFTs for EIP-721
author: Jeremy Z (@wnft)
discussions-to: https://ethereum-magicians.org/t/eip-5496-multi-privilege-management-extension-for-erc-721/10427

> status: Last Call

Reviewer: You should rebase this PR against master.

Author: done

type: Standards Track
category: ERC

> created: 2022-7-30

Reviewer 2: This will probably be one of the most valuable expandable use case I could suggest: Consider mentioning how with EIP-1271 this EIP-5496 could be useful to authorize a privilege with smart contract, ideally providing an reference implementation. If you don't know how to do it, feel free to ask question in the [etheruem-magicians.org thread](https://ethereum-magicians.org/t/eip-5496-multi-privilege-management-extension-for-erc-721/10427) thread

Author: We will add new description which for Compliant contract, it is RECOMMENDED to use EIP-1271 to validate the signatures.

description: Create shareable multi-privilege NFTs for EIP-721
author: Jeremy Z (@wnft)
discussions-to: https://ethereum-magicians.org/t/eip-5496-multi-privilege-management-extension-for-erc-721/10427
status: Last Call

> last-call-deadline: 2022-11-14

Reviewer: ```suggestion
last-call-deadline: 2022-11-29
```

## Abstract

> This EIP defines an interface extending [EIP-721](./eip-721.md) to provide shareable multi-privileges for NFTs. Privileges may be on-chain (voting rights, permission to claim an airdrop) or off-chain (a coupon for an online store, a discount at a local restaurant, access to VIP lounges in airports). Each NFT may contain many privileges, and the holder of a privilege can verifiably transfer that privilege to others. Privileges may be non-shareable or shareable. Shareable privileges can be cloned, with the provider able to adjust the details according to the spreading path. Expiration periods can also be set for each privilege.

Reviewer 2: My first question would be, would it need to restrict to 721 or does it also work for EIP-1155? It would help adoption to support both but also up to you.

Author: We will not support EIP-1155.
In EIP-5496 each privilege links to one token id of EIP-721, due to EIP-721 is non-fungible and unique. It's difficult to manage privilege for each token id in EIP-1155 which may contain multiple fungible NFTs with the same id.

## Motivation

> Many NFTs have functions other than just being used as profile pictures or art collections, they may have real utilities in different scenarios. For example, a fashion store may give a discount for its own NFT holders; a DAO member NFT holder can vote for the proposal of how to use their treasury; a dApp may create an airdrop event to attract a certain group of people like some blue chip NFT holders to claim; Walmart can issue its membership card on chain (as an NFT) and give certain privileges when the members shop at Walmart stores, etc. There are cases when people who own NFTs do not necessarily want to use their privileges. By providing additional data recording different privileges a NFT collection has and interfaces to manage them, users can transfer or sell privileges without losing their ownership of the NFT.

Reviewer: Something that might be useful to include is why we should use EIP-5496 privileges over representing privileges **as** NFTs themselves.

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> Every contract compliant to the `ERC5496` extension MUST implement the `IERC5496` interface. The **shareable multi-privilege extension** is OPTIONAL for EIP-721 contracts.

Reviewer: ```suggestion
Every contract complying with this standard MUST implement the `IERC5496` interface. The **shareable multi-privilege extension** is OPTIONAL for EIP-721 contracts.
```

Author: I updated the markdown file and pls review it

> 

Reviewer 2: Line 49 seems a copy-pasting error, please fix and check instances else where.

Reviewer 3: done

Author: done

    /// @param tokenId The identifier of the queried NFT
    /// @param privilegeId The identifier of the queried privilege

>     /// @param referrer The address of the referrer

Reviewer 2: could you elaborate of what a referrer is in this context?

Author: For example, a local pizza shop offers a 30% off Coupon and the owner of the shop encourages their consumers to share the coupon with friends, then the friends can get the coupon. Let's say Tom gets 30% off Coupon from the shop and he shares the coupon with Alice. Alice gets the coupon too and Alice's referrer is Tom.  For some certain cases, Tom may get more rewards from the shop. This's a good way to expand markets for merchants.

Reviewer: This is good content! It should go into the EIP itself.

Author: done

>     /// @return Whether the operation is successful or not

Reviewer 2: Could you aslo specify what will happen to the "privileges" if a token is transfered?

Author: If the owner of the EIP-5496 NFT transfer ownership to another user, there is no impact on "privileges". But errors may be occurred when the owner tries to withdraw the original EIP-721 token from EIP-5496 via calling function unwrap() in our implementation contract if any available privilege is still ongoing. We protect the rights of holders of the privileges by checking the last expire date of the privilege.
  
function unwrap(uint256 tokenId, address to) external {
        require(getBlockTimestamp() >= privilegeBook[tokenId].lastExpiresAt, "privilege not yet expired");
        require(ownerOf(tokenId) == msg.sender, "not owner");
        _burn(tokenId);
        IERC721(nft).transferFrom(address(this), to, tokenId);
        emit Unwrap(nft, tokenId, msg.sender, to);
    }

EIP-721 only records the ownership and its transfer, the privileges of an NFT are not recorded on-chain. This extension would allow merchants/projects to give out a certain privilege to a specified group of people, and owners of the privileges can manage each one of the privileges independently. This facilitates a great possibility for NFTs to have real usefulness. 

> ### Shareable privileges

Reviewer: ```suggestion
### Shareable Privileges
```

> The number of privilege holders is limited by the number of NFTs if privileges are non-shareable. A shareable privilege means the original privilege holder can copy the privilege and give it to the others, not transferring his/her own privilege to the others. This mechanism greatly enhances the spread of privileges as well as the adoption of NFTs. 

Reviewer: ```suggestion
The number of privilege holders is limited by the number of NFTs if privileges are non-shareable. A shareable privilege means the original privilege holder can copy the privilege and give it to others, not transferring his/her own privilege to them. This mechanism greatly enhances the spread of privileges as well as the adoption of NFTs. 
```

Author: done

### Shareable Privileges

The number of privilege holders is limited by the number of NFTs if privileges are non-shareable. A shareable privilege means the original privilege holder can copy the privilege and give it to others, not transferring his/her own privilege to them. This mechanism greatly enhances the spread of privileges as well as the adoption of NFTs.

### Expire Date Type

The expiry timestamp of a privilege is a timestamp and stored in `uint256` typed variables.

### Beneficiary of Referrer

For example, a local pizza shop offers a 30% off Coupon and the owner of the shop encourages their consumers to share the coupon with friends, then the friends can get the coupon. Let's say Tom gets 30% off Coupon from the shop and he shares the coupon with Alice. Alice gets the coupon too and Alice's referrer is Tom. For some certain cases, Tom may get more rewards from the shop. This will help the merchants by spreading the promotion information to more consumers.

### Proposal: NFT Transfer

> If the owner of this standard NFT transfer ownership to another user, there is no impact on "privileges". But errors may be occurred when the owner tries to withdraw the original [EIP-721](./eip-721.md) token from [EIP-5496](./eip-5496.md) via calling function `unwrap()` in our implementation contract if any available privilege is still ongoing. We protect the rights of holders of the privileges by checking the last expire date of the privilege.

Reviewer: ```suggestion
If the owner of this standard NFT transfers ownership to another user, there is no impact on "privileges". But errors may occur when the owner tries to withdraw the original [EIP-721](./eip-721.md) token from a wrapper via calling function `unwrap()` in our implementation contract if any available privilege is still valid. We protect the rights of holders of the privileges by checking the last expire date of the privilege.
```
